File: scripts/update-library.py
#!/usr/bin/env python3

"""
Virtual Bookshelf Library - Python Update Script

This script fetches book data from Google Books API and updates the library.
It can be run manually or scheduled to run daily via cron job.

Usage: python scripts/update-library.py

Requirements: Python 3.6+ (no additional packages required)
"""

import json
import os
import sys
import time
import urllib.request
import urllib.parse
import urllib.error
from datetime import datetime, timedelta
import re

class LibraryUpdater:
    def __init__(self):
        self.script_dir = os.path.dirname(os.path.abspath(__file__))
        self.project_dir = os.path.dirname(self.script_dir)
        self.data_path = os.path.join(self.project_dir, 'data', 'books.json')
        self.covers_path = os.path.join(self.project_dir, 'assets', 'covers')
        self.api_key = os.environ.get('GOOGLE_BOOKS_API_KEY', '')
        self.rate_limit_delay = 1.0  # 1 second between API calls
        
        # Ensure covers directory exists
        os.makedirs(self.covers_path, exist_ok=True)

    def run(self):
        print('🚀 Starting library update...')
        
        try:
            books = self.load_books()
            print(f'📚 Found {len(books)} books to process')
            
            updated_books = self.update_books_data(books)
            self.save_books(updated_books)
            
            print('✅ Library update completed successfully!')
            print(f'📊 Updated {len(updated_books)} books')
            
        except Exception as error:
            print(f'❌ Error updating library: {error}')
            sys.exit(1)

    def load_books(self):
        try:
            with open(self.data_path, 'r', encoding='utf-8') as file:
                data = json.load(file)
                return data.get('books', [])
        except FileNotFoundError:
            print('📝 No books.json found, creating new library')
            return []
        except json.JSONDecodeError as error:
            raise Exception(f'Invalid JSON in books.json: {error}')

    def save_books(self, books):
        data = {
            'books': books,
            'lastUpdated': datetime.now().isoformat()
        }
        
        # Ensure data directory exists
        os.makedirs(os.path.dirname(self.data_path), exist_ok=True)
        
        with open(self.data_path, 'w', encoding='utf-8') as file:
            json.dump(data, file, indent=2, ensure_ascii=False)
        
        print(f'💾 Saved {len(books)} books to {self.data_path}')

    def update_books_data(self, books):
        updated_books = []
        
        for i, book in enumerate(books):
            print(f'🔄 Processing book {i + 1}/{len(books)}: {book.get("title", book.get("isbn", "Unknown"))}')
            
            try:
                updated_book = self.fetch_book_data(book)
                updated_books.append(updated_book)
                
                # Rate limiting
                if i < len(books) - 1:
                    time.sleep(self.rate_limit_delay)
                    
            except Exception as error:
                print(f'⚠️  Failed to update book {book.get("isbn", "unknown")}: {error}')
                # Keep the original book data if API fails
                updated_books.append(book)
        
        return updated_books

    def fetch_book_data(self, book):
        # If book already has complete data and was updated recently, skip API call
        if self.is_book_data_complete(book) and self.is_recently_updated(book):
            print(f'⏭️  Skipping {book.get("title", "Unknown")} (recently updated)')
            return book

        api_data = None
        
        # Try to fetch by ISBN first
        isbn = book.get('isbn', '')
        if isbn and (isbn.startswith('978') or isbn.startswith('979')):
            api_data = self.fetch_from_google_books(f'isbn:{isbn}')
        
        # If ISBN search failed, try title + author
        if not api_data and book.get('title') and book.get('author'):
            query = f'intitle:"{book["title"]}" inauthor:"{book["author"]}"'
            api_data = self.fetch_from_google_books(query)
        
        # Merge API data with existing book data
        updated_book = book.copy()
        
        if api_data:
            volume_info = api_data.get('volumeInfo', {})
            
            # Update fields only if they're missing or empty
            if not updated_book.get('title') and volume_info.get('title'):
                updated_book['title'] = volume_info['title']
            
            if not updated_book.get('author') and volume_info.get('authors'):
                updated_book['author'] = ', '.join(volume_info['authors'])
            
            if not updated_book.get('synopsis') and volume_info.get('description'):
                updated_book['synopsis'] = self.clean_description(volume_info['description'])
            
            if not updated_book.get('genre') and volume_info.get('categories'):
                updated_book['genre'] = volume_info['categories'][0]
            
            if not updated_book.get('pageCount') and volume_info.get('pageCount'):
                updated_book['pageCount'] = volume_info['pageCount']
            
            if not updated_book.get('averageRating') and volume_info.get('averageRating'):
                updated_book['averageRating'] = volume_info['averageRating']
            
            # Update cover image
            image_links = volume_info.get('imageLinks', {})
            if image_links:
                cover_url = (image_links.get('large') or 
                           image_links.get('medium') or 
                           image_links.get('small') or 
                           image_links.get('thumbnail'))
                
                if cover_url:
                    updated_book['coverUrl'] = cover_url.replace('http:', 'https:')
                    
                    # Optionally download and cache the cover image
                    try:
                        self.download_cover(updated_book['isbn'], updated_book['coverUrl'])
                    except Exception as error:
                        print(f'⚠️  Failed to download cover for {book.get("title", "Unknown")}')
            
            updated_book['lastApiUpdate'] = datetime.now().isoformat()
            print(f'✅ Updated {updated_book.get("title", "Unknown")}')
        else:
            print(f'ℹ️  No API data found for {book.get("title", book.get("isbn", "Unknown"))}')
        
        return updated_book

    def fetch_from_google_books(self, query):
        encoded_query = urllib.parse.quote(query)
        api_url = f'https://www.googleapis.com/books/v1/volumes?q={encoded_query}'
        
        if self.api_key:
            api_url += f'&key={self.api_key}'
        
        try:
            with urllib.request.urlopen(api_url, timeout=10) as response:
                data = json.loads(response.read().decode('utf-8'))
                
                if data.get('items') and len(data['items']) > 0:
                    return data['items'][0]  # Return the first match
                
                return None
                
        except urllib.error.HTTPError as error:
            raise Exception(f'Google Books API HTTP error: {error.code}')
        except urllib.error.URLError as error:
            raise Exception(f'Google Books API URL error: {error.reason}')
        except json.JSONDecodeError as error:
            raise Exception(f'Google Books API JSON error: {error}')

    def download_cover(self, isbn, cover_url):
        filename = f'{isbn}.jpg'
        filepath = os.path.join(self.covers_path, filename)
        
        # Check if cover already exists
        if os.path.exists(filepath):
            return  # Cover already exists
        
        try:
            with urllib.request.urlopen(cover_url, timeout=10) as response:
                with open(filepath, 'wb') as file:
                    file.write(response.read())
            print(f'📸 Downloaded cover for {isbn}')
        except Exception as error:
            raise Exception(f'Failed to download cover: {error}')

    def clean_description(self, description):
        """Remove HTML tags and clean up the description"""
        # Remove HTML tags
        clean_text = re.sub(r'<[^>]*>', '', description)
        
        # Replace HTML entities
        clean_text = clean_text.replace('&quot;', '"')
        clean_text = clean_text.replace('&amp;', '&')
        clean_text = clean_text.replace('&lt;', '<')
        clean_text = clean_text.replace('&gt;', '>')
        
        return clean_text.strip()

    def is_book_data_complete(self, book):
        required_fields = ['title', 'author', 'synopsis', 'coverUrl', 'pageCount']
        return all(book.get(field) for field in required_fields)

    def is_recently_updated(self, book):
        last_update = book.get('lastApiUpdate')
        if not last_update:
            return False
        
        try:
            last_update_date = datetime.fromisoformat(last_update.replace('Z', '+00:00'))
            days_since_update = (datetime.now() - last_update_date).days
            return days_since_update < 7  # Consider recent if updated within 7 days
        except (ValueError, TypeError):
            return False

def main():
    """Main entry point when script is run directly"""
    updater = LibraryUpdater()
    updater.run()

if __name__ == '__main__':
    main()

File: scripts/update-library.js
#!/usr/bin/env node

/**
 * Virtual Bookshelf Library - Update Script
 * 
 * This script fetches book data from Google Books API and updates the library.
 * It can be run manually or scheduled to run daily via cron job.
 * 
 * Usage: node scripts/update-library.js
 */

const fs = require('fs').promises;
const path = require('path');
const https = require('https');

class LibraryUpdater {
    constructor() {
        this.dataPath = path.join(__dirname, '..', 'data', 'books.json');
        this.coversPath = path.join(__dirname, '..', 'assets', 'covers');
        this.apiKey = process.env.GOOGLE_BOOKS_API_KEY || ''; // Optional API key for higher rate limits
        this.rateLimitDelay = 1000; // 1 second between API calls
    }

    async run() {
        console.log('🚀 Starting library update...');
        
        try {
            const books = await this.loadBooks();
            console.log(`📚 Found ${books.length} books to process`);
            
            const updatedBooks = await this.updateBooksData(books);
            await this.saveBooks(updatedBooks);
            
            console.log('✅ Library update completed successfully!');
            console.log(`📊 Updated ${updatedBooks.length} books`);
            
        } catch (error) {
            console.error('❌ Error updating library:', error.message);
            process.exit(1);
        }
    }

    async loadBooks() {
        try {
            const data = await fs.readFile(this.dataPath, 'utf8');
            const parsed = JSON.parse(data);
            return parsed.books || [];
        } catch (error) {
            if (error.code === 'ENOENT') {
                console.log('📝 No books.json found, creating new library');
                return [];
            }
            throw error;
        }
    }

    async saveBooks(books) {
        const data = {
            books: books,
            lastUpdated: new Date().toISOString()
        };
        
        await fs.writeFile(this.dataPath, JSON.stringify(data, null, 2));
        console.log(`💾 Saved ${books.length} books to ${this.dataPath}`);
    }

    async updateBooksData(books) {
        const updatedBooks = [];
        
        for (let i = 0; i < books.length; i++) {
            const book = books[i];
            console.log(`🔄 Processing book ${i + 1}/${books.length}: ${book.title || book.isbn}`);
            
            try {
                const updatedBook = await this.fetchBookData(book);
                updatedBooks.push(updatedBook);
                
                // Rate limiting
                if (i < books.length - 1) {
                    await this.delay(this.rateLimitDelay);
                }
                
            } catch (error) {
                console.warn(`⚠️  Failed to update book ${book.isbn}: ${error.message}`);
                // Keep the original book data if API fails
                updatedBooks.push(book);
            }
        }
        
        return updatedBooks;
    }

    async fetchBookData(book) {
        // If book already has complete data and was updated recently, skip API call
        if (this.isBookDataComplete(book) && this.isRecentlyUpdated(book)) {
            console.log(`⏭️  Skipping ${book.title} (recently updated)`);
            return book;
        }

        let apiData = null;
        
        // Try to fetch by ISBN first
        if (book.isbn && (book.isbn.startsWith('978') || book.isbn.startsWith('979'))) {
            apiData = await this.fetchFromGoogleBooks(`isbn:${book.isbn}`);
        }
        
        // If ISBN search failed, try title + author
        if (!apiData && book.title && book.author) {
            const query = `intitle:"${book.title}" inauthor:"${book.author}"`;
            apiData = await this.fetchFromGoogleBooks(query);
        }
        
        // Merge API data with existing book data
        const updatedBook = { ...book };
        
        if (apiData) {
            const volumeInfo = apiData.volumeInfo;
            
            // Update fields only if they're missing or empty
            if (!updatedBook.title && volumeInfo.title) {
                updatedBook.title = volumeInfo.title;
            }
            
            if (!updatedBook.author && volumeInfo.authors) {
                updatedBook.author = volumeInfo.authors.join(', ');
            }
            
            if (!updatedBook.synopsis && volumeInfo.description) {
                updatedBook.synopsis = this.cleanDescription(volumeInfo.description);
            }
            
            if (!updatedBook.genre && volumeInfo.categories) {
                updatedBook.genre = volumeInfo.categories[0];
            }
            
            if (!updatedBook.pageCount && volumeInfo.pageCount) {
                updatedBook.pageCount = volumeInfo.pageCount;
            }
            
            if (!updatedBook.averageRating && volumeInfo.averageRating) {
                updatedBook.averageRating = volumeInfo.averageRating;
            }
            
            // Update cover image
            if (volumeInfo.imageLinks) {
                const coverUrl = volumeInfo.imageLinks.large || 
                               volumeInfo.imageLinks.medium || 
                               volumeInfo.imageLinks.small || 
                               volumeInfo.imageLinks.thumbnail;
                
                if (coverUrl) {
                    updatedBook.coverUrl = coverUrl.replace('http:', 'https:');
                    
                    // Optionally download and cache the cover image
                    try {
                        await this.downloadCover(updatedBook.isbn, updatedBook.coverUrl);
                    } catch (error) {
                        console.warn(`⚠️  Failed to download cover for ${book.title}`);
                    }
                }
            }
            
            updatedBook.lastApiUpdate = new Date().toISOString();
            console.log(`✅ Updated ${updatedBook.title}`);
        } else {
            console.log(`ℹ️  No API data found for ${book.title || book.isbn}`);
        }
        
        return updatedBook;
    }

    async fetchFromGoogleBooks(query) {
        const encodedQuery = encodeURIComponent(query);
        const apiUrl = `https://www.googleapis.com/books/v1/volumes?q=${encodedQuery}${this.apiKey ? `&key=${this.apiKey}` : ''}`;
        
        try {
            const response = await this.httpGet(apiUrl);
            const data = JSON.parse(response);
            
            if (data.items && data.items.length > 0) {
                return data.items[0]; // Return the first match
            }
            
            return null;
        } catch (error) {
            throw new Error(`Google Books API error: ${error.message}`);
        }
    }

    async downloadCover(isbn, coverUrl) {
        const filename = `${isbn}.jpg`;
        const filepath = path.join(this.coversPath, filename);
        
        // Check if cover already exists
        try {
            await fs.access(filepath);
            return; // Cover already exists
        } catch (error) {
            // Cover doesn't exist, download it
        }
        
        return new Promise((resolve, reject) => {
            const file = require('fs').createWriteStream(filepath);
            
            https.get(coverUrl, (response) => {
                if (response.statusCode === 200) {
                    response.pipe(file);
                    file.on('finish', () => {
                        file.close();
                        console.log(`📸 Downloaded cover for ${isbn}`);
                        resolve();
                    });
                } else {
                    reject(new Error(`HTTP ${response.statusCode}`));
                }
            }).on('error', (error) => {
                reject(error);
            });
        });
    }

    httpGet(url) {
        return new Promise((resolve, reject) => {
            https.get(url, (response) => {
                let data = '';
                
                response.on('data', (chunk) => {
                    data += chunk;
                });
                
                response.on('end', () => {
                    if (response.statusCode === 200) {
                        resolve(data);
                    } else {
                        reject(new Error(`HTTP ${response.statusCode}: ${data}`));
                    }
                });
            }).on('error', (error) => {
                reject(error);
            });
        });
    }

    cleanDescription(description) {
        // Remove HTML tags and clean up the description
        return description
            .replace(/<[^>]*>/g, '') // Remove HTML tags
            .replace(/&quot;/g, '"')
            .replace(/&amp;/g, '&')
            .replace(/&lt;/g, '<')
            .replace(/&gt;/g, '>')
            .trim();
    }

    isBookDataComplete(book) {
        return book.title && book.author && book.synopsis && book.coverUrl && book.pageCount;
    }

    isRecentlyUpdated(book) {
        if (!book.lastApiUpdate) return false;
        
        const lastUpdate = new Date(book.lastApiUpdate);
        const daysSinceUpdate = (Date.now() - lastUpdate.getTime()) / (1000 * 60 * 60 * 24);
        
        return daysSinceUpdate < 7; // Consider recent if updated within 7 days
    }

    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// Run the updater if this script is executed directly
if (require.main === module) {
    const updater = new LibraryUpdater();
    updater.run();
}

module.exports = LibraryUpdater;
